bootcommander: init at 1.17.00, libopenblt: init at 1.17.00

simoneruffini opened the pull request:
###### Description of changes

Added bootcommander flasher utility and it's dynamic library dependency libopenblt

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2456

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+{ lib, stdenv, fetchgit, cmake, libopenblt }:
+
+stdenv.mkDerivation rec {
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+  pname = "bootcommander";
+  version = "011500";
+
+  src = fetchgit {
```
Please use fetchFromGitHub

simoneruffini replied:
the github repo is a mirror

simoneruffini replied:
does fetchFromGithub support partial clone? because the repo is too big ad fool of useless examples

SuperSandro2000 replied:
it does support sparse checkout

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+stdenv.mkDerivation rec {
+
+  pname = "bootcommander";
+  version = "011500";
```
You are sure the version number isn't 1.15.0?

simoneruffini replied:
the relase page states no version number but a tag called `libopenblt_011500`
the problem is that this is not the actual version of the application but the version of the repo that contains both the library and the bootcommander application. Therefore I don't know how to set the correct version of the application and library and use that variable to fetch the repo since it differs to the github tag

SuperSandro2000 replied:
and that version in the tag does not translate to anything? then we just use that, because it is easier

fgaz replied:
Seems like official versions use dots: https://sourceforge.net/projects/openblt/files/OpenBLT%20stable/

simoneruffini replied:
on svn/sourceforge yes, but on github no, so you would need a different value for pulling from GH and one for assigning the version

fgaz replied:
To me it looks like versions and tags correspond:

![image](https://github.com/NixOS/nixpkgs/assets/8182846/83c79712-87bc-40cf-ae61-c30b3a746663)
![image](https://github.com/NixOS/nixpkgs/assets/8182846/88d375be-9e6b-44da-932a-8f05ae252c44)

so this should work:

```nix
version = "1.16.00";
src = fetchFromGitHub {
  rev = "openblt_v0${lib.replaceStrings ["."] [""] version}";
};
```

Thechincally this will break as soon as version 10.x.y is released, but I think it's good enough (I don't see any padding function in lib/)

simoneruffini replied:
Is this really necessary? Since we already know that it will break and the value by itself is not very usefull because the software has a different version, can't we keep it simple like this? Moreover will automatic updates handled by ryantm support this?

fgaz replied:
> Is this really necessary?

Yes, the `version` attribute should reflect the actual version. Tools such as repology rely on it. There are [many cases like this one](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22rev+%3D%22+replaceStrings&type=code).

> Since we already know that it will break

It will break... in 10 major versions, which won't happen any time soon. You can make it more robust by conditionally adding that zero if you'd like to.

> Moreover will automatic updates handled by ryantm support this?

I think so. It can get notified by repology and the revision will is derived from the version.

fgaz replied:
https://nixos.org/manual/nixpkgs/stable/index.html#var-passthru-updateScript

simoneruffini replied:
done

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+    sha256 = "sha256-cWdc6Imd6IBqYQRNEafp3bDDQLgdpE90fxSkY+vzhAM=";
+  };
+
+  preConfigure = "cd ./Host/Source/BootCommander/";
```
We should use sourceRoot or similar

simoneruffini replied:
I can't use sourceroot because `make` saves the binary output in different locations that otherwise would not be accessible with sourceroot.

simoneruffini replied:
I tried over and over to get this thing to work and I can't. I need help @SuperSandro2000 @fgaz 
Since the CMakeList.txt file uses fixed paths for the executables ouptput dirs (https://github.com/feaser/openblt/blob/0e1104d711c3993cf83f47f7a6bef3485efeb3dc/Host/Source/LibOpenBLT/CMakeLists.txt#L71) I can't use `sourceRoot="source/Host/Source/BootCommander"` otherwise the linker fails with error. The reason is that the compilation works but the linker tries to output the file to the directory `../../../` that is out of scope in the nix sandbox build dir. The same values that control this behaviour (`PROJECT_OUTPUT_DIRECTORY,CMAKE_RUNTIME_OUTPUT_DIRECTORY`...) can't be overridden in `cmakeFlags` because they will not be read since cmake has precedence on the actual ones defined in the CMakeList.txt instead of the command line ones. So the last options are either to patch the cmake file or to just cd in to the folder and build from there. Otherwise I don't know how to build this program.

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+  preConfigure = "cd ./Host/Source/BootCommander/";
+
+  installPhase = ''
+    ls
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  installPhase = ''
+    ls
+    cd $NIX_BUILD_TOP/openblt/Host/
```
Please use relative directors

simoneruffini replied:
In this case relative directories don't give  context where files are and it is more difficult to debug and work later on, the relative path should be something like `../../../../`

SuperSandro2000 replied:
Using NIX_BUILD_TOP should be avoided as much as possible as it doesn't need to be /build, eg on darwin it is something else.

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+  installPhase = ''
+    ls
+    cd $NIX_BUILD_TOP/openblt/Host/
+    install -D BootCommander $out/bin/BootCommander
```
You are sure this isn't supposed to be lowercased?

simoneruffini replied:
the binary output unfortunately is camel case

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+    install -D BootCommander $out/bin/BootCommander
+  '';
+
+  nativeBuildInputs = [ cmake libopenblt ];
```
libopenblt hoes to buildInputs most likely

simoneruffini replied:
true!

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+  nativeBuildInputs = [ cmake libopenblt ];
+
+  meta = with lib; {
+    description = "BootCommander is a CLI program for performing firmware updates on a microcontroller that runs the OpenBLT bootloader";
```
```suggestion
    description = "CLI program to performe firmware updates on microcontrollers that run the OpenBLT bootloader";
```

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,32 @@
+  meta = with lib; {
+    description = "BootCommander is a CLI program for performing firmware updates on a microcontroller that runs the OpenBLT bootloader";
+    homepage = "https://www.feaser.com/openblt/doku.php?id=manual:bootcommander";
+    license = licenses.gpl3;
```
Is that gpl3Only or gpl3Plus?

simoneruffini replied:
gpl3 only

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ lib, stdenv, fetchgit, cmake, libusb1 }:
+
+stdenv.mkDerivation rec {
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+stdenv.mkDerivation rec {
+
+  pname = "libopenblt";
+  version = "011500";
```
Same question as above

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+  pname = "libopenblt";
+  version = "011500";
+
+  src = fetchgit {
```
Use fetchFromGitHub

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+    install -D openblt.h $out/include/openblt.h
+  '';
+
+  nativeBuildInputs = [ cmake libusb1 ];
```
libusb1 goes to buildInputs

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+  nativeBuildInputs = [ cmake libusb1 ];
+
+  meta = with lib; {
+    description = "The OpenBLT Host Library contains an API for communicating with a microcontroller running the OpenBLT bootloader";
```
```suggestion
    description = "API for communicating with a microcontrollers running the OpenBLT bootloader";
```
See contributing doc, we don't repeat the package/library name

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+  installPhase = ''
+    cd $NIX_BUILD_TOP/openblt/Host/
+    install -D libopenblt.so $out/lib/libopenblt.so
+    cd $NIX_BUILD_TOP/openblt/Host/Source/LibOpenBLT/
```
We can give install the full relative path and avoid the cd

simoneruffini replied:
for the relative paths as above, the cd can be removed

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+  preConfigure = "cd ./Host/Source/LibOpenBLT/";
+
+  installPhase = ''
+    cd $NIX_BUILD_TOP/openblt/Host/
```
Same as above

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "sha256-cWdc6Imd6IBqYQRNEafp3bDDQLgdpE90fxSkY+vzhAM=";
+  };
+
+  preConfigure = "cd ./Host/Source/LibOpenBLT/";
```
Same as above

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+    url = "https://github.com/feaser/openblt.git";
+    rev = "openblt_v${version}";
+    sparseCheckout = [ "Host/Source/" ];
+    sha256 = "sha256-cWdc6Imd6IBqYQRNEafp3bDDQLgdpE90fxSkY+vzhAM=";
```
```suggestion
    hash = "sha256-cWdc6Imd6IBqYQRNEafp3bDDQLgdpE90fxSkY+vzhAM=";
```

fgaz reviewed (changes requested):
Please split into 2 separate commits: libopenblt first, then bootcommander

simoneruffini committed:
libopenblt: init at 1.17.00

simoneruffini committed:
bootcommander: init at 1.17.00

Update pkgs/by-name/bo/bootcommander/package.nix

Co-authored-by: Arne Keller <2012gdwu+github@posteo.de>

Update pkgs/by-name/li/libopenblt/package.nix

Co-authored-by: Arne Keller <2012gdwu+github@posteo.de>

update Bootcommander package.nix

commit

simoneruffini commented:
Any updates?

SuperSandro2000 commented on pkgs/development/libraries/libopenblt/default.nix:
```diff
@@ -0,0 +1,33 @@
+  preConfigure = "cd ./Host/Source/LibOpenBLT/";
+
+  installPhase = ''
+    install -D $NIX_BUILD_TOP/source/Host/libopenblt.so $out/lib/libopenblt.so
+    install -D $NIX_BUILD_TOP/source/Host/Source/LibOpenBLT/openblt.h $out/include/openblt.h
```
We can rely on pwd here

```suggestion
    install -D Host/libopenblt.so $out/lib/libopenblt.so
    install -D Host/Source/LibOpenBLT/openblt.h $out/include/openblt.h
```

simoneruffini replied:
tried, doesn't work

simoneruffini replied:
During installphase I'm here `/build/source/Host/Source/LibOpenBLT/build` but I need to take my file from here `/build/source/Host/`. I would like to be in the `/build/source` folder so I can correctly type out the path location and don't use something not understandable like `../../../libopenblt.so`

simoneruffini replied:
Check the solution on the last commit

SuperSandro2000 commented on pkgs/development/embedded/bootcommander/default.nix:
```diff
@@ -0,0 +1,31 @@
+  preConfigure = "cd ./Host/Source/BootCommander/";
+
+  installPhase = ''
+    install -D $NIX_BUILD_TOP/source/Host/BootCommander $out/bin/BootCommander
```
```suggestion
    install -D Host/BootCommander $out/bin/BootCommander
```

simoneruffini commented:
Do libraries need to move from `pkgs/development/libaries` to `pkgs/by-name/li/` too?

FliegendeWurst commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,40 @@
+    cd ./Host/Source/BootCommander/
+  '';
+
+  installPhase = ''
+    cd $SRCDIRPWD
+    install -D "Host/BootCommander" "$out/bin/BootCommander"
+    ln -s "$out/bin/BootCommander" "$out/bin/bootcommander"
+  '';
```
```suggestion
  installPhase = ''
    cd $SRCDIRPWD
    runHook preInstall
    install -D "Host/BootCommander" "$out/bin/BootCommander"
    ln -s "$out/bin/BootCommander" "$out/bin/bootcommander"
    runHook postInstall
  '';
```

FliegendeWurst commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,40 @@
+  buildInputs = [ libopenblt ];
+  nativeBuildInputs = [ cmake ];
+
+  meta = {
```
Add meta.mainProgram please

FliegendeWurst commented on pkgs/by-name/li/libopenblt/package.nix:
```diff
@@ -0,0 +1,40 @@
+    cd ./Host/Source/LibOpenBLT/
+  '';
+
+  installPhase = ''
+    cd $srcDirPwd
+    install -D "Host/libopenblt.so" "$out/lib/libopenblt.so"
+    install -D "Host/Source/LibOpenBLT/openblt.h" "$out/include/openblt.h"
+  '';
```
```suggestion
  installPhase = ''
    cd $srcDirPwd
    runHook preInstall
    install -D "Host/libopenblt.so" "$out/lib/libopenblt.so"
    install -D "Host/Source/LibOpenBLT/openblt.h" "$out/include/openblt.h"
    runHook postInstall
  '';
```

FliegendeWurst commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,49 @@
+  src = fetchFromGitHub {
+    owner = "feaser";
+    repo = "openblt";
+    rev = "openblt_v${lib.replaceStrings [ "." ] [ "" ] version}"; # will break at version 10.x.x
```
```suggestion
    rev = "openblt_v0${lib.replaceStrings [ "." ] [ "" ] version}"; # will break at version 10.x.x
```

fgaz commented on pkgs/by-name/li/libopenblt/package.nix:
Do this in the previous commit

fgaz replied:
GH's review interface is as awful as always and it doesn't show that I commented on the openblt reformat you did on the _second_ commit

fgaz commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,49 @@
+  libopenblt,
+}:
+
+let
+  pname = "bootcommander";
+  version = "1.15.00";
+in
```
put this inside mkDerivation. it's `rec` anyway

fgaz commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,49 @@
+  pname = "bootcommander";
+  version = "1.15.00";
+in
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

fgaz commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,49 @@
+  src = fetchFromGitHub {
+    owner = "feaser";
+    repo = "openblt";
+    rev = "openblt_v0${lib.replaceStrings [ "." ] [ "" ] version}"; # will break at version 10.x.x
```
```suggestion
    rev = "refs/tags/openblt_v0${lib.replaceStrings [ "." ] [ "" ] finalAttrs.version}"; # will break at version 10.x.x
```

simoneruffini replied:
why? its working fine without that?

fgaz replied:
To disambiguate from branches and other refs

fgaz reviewed (changes requested):
Same in the other file

phanirithvij commented on pkgs/by-name/bo/bootcommander/package.nix:
```diff
@@ -0,0 +1,45 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "bootcommander";
+  version = "1.15.00";
```
title says 1.7.00. also currently the latest tag is [openblt_v011901](https://github.com/feaser/openblt/releases/tag/openblt_v011901)
